--- xerces-c-src_2_7_0/src/xercesc/runConfigure	2005-09-07 08:55:53.000000000 -0700
+++ xerces-c-src_2_7_0_gm/src/xercesc/runConfigure	2006-02-01 20:40:57.000000000 -0700
@@ -645,7 +645,7 @@
       CXXVER="v5compat"
       ;;
       
-   xlC* | xlc* | g++ | c++ | cc | CC | aCC | icc | icpc | ICC | cxx | ecc | QCC )
+   xlC* | xlc* | g++ | c++ | cc | CC | aCC | icc | icpc | ICC | cxx | ecc | QCC | arm-linux-g++)
       CXX="$cppcompiler"
       ;;
     
